### PR TITLE
Gradient colormap

### LINKED ISSUE #306: Scatterplot: add gradient colormap for continuous overlay variables
------------------
***NOTE: This is now a duplicate of https://github.com/VEuPathDB/web-eda/issues/1399***
------------------

## Tasks

- [x] Pick sequential gradient map. Decided on the darkest 90% of colors from [oslo colormap](https://www.fabiocrameri.ch/colourmaps/). 
- [x] Gradient colormap legend.
- [x] Handle how no data interacts with gradient colormap. Plan: put the No Data in it's own, non-clickable, slot in the legend. Add prop to send hasMissingData to colormap legend and it can handle adding a no data line.
- [x] plot.data handle missing data. Needs to return a new seriesGradientColormap named "no data" or similar.
- [x] data service updates. 
- [x] web-eda general implementation
- [x] web-eda gradient colorscale turns off trendlines
- [x] Use annotated displayRangeMin and max if given (look at `defaultDependentAxisRange`). If not given, use rangeMin and rangeMax.
- [x] web-eda handle the dual overlay var constraints.

## Associated PRs
- web-components [#275](https://github.com/VEuPathDB/web-components/pull/275)
- plot.data [PR 124](https://github.com/VEuPathDB/plot.data/pull/124)
- web-eda [PR #790](https://github.com/VEuPathDB/web-eda/pull/790)
- data service [PR #138](https://github.com/VEuPathDB/EdaDataService/pull/138)

## Open questions

- [ ] Markers? Filled circles?
- [ ] Pick diverging map. IP
- [ ] How to handle "discrete" vars? If we let integers with < 9 unique values in, do they get categorical, gradient, or discrete? 
- [ ] Var constraints more generally. Right now set up to use var info to determine how the viz runs (web-eda). Are we happy sending all numeric (for example) vars to go through series gradient colormap? If var has values = 1:7, they could still get a seriesGradientColormap but then when sent to the legend are legend-ized appropriately.

## To Dos for later cards

- [ ] Allow dates
- [ ] Add controls for user to select colormap, range, midpoint (maybe future phase?)

### PR DESCRIPTION
**Pull request by asizemore**
With plot.data [PR #124](https://github.com/VEuPathDB/plot.data/pull/124), data service [PR #138](https://github.com/VEuPathDB/EdaDataService/pull/138), and web-components [PR #275](https://github.com/VEuPathDB/web-components/pull/275), resolves #306.

When the scatter response contains a 'seriesGradientColorscale' column, we now map that number to a color using the appropriate colormap.

Note that we have a bit of a var constraint mess to sort out. If using a gradient colorscale, then we want to disable trend lines. Currently here I'm using dataType. plot.data is using dataShape=continuous. Not a terrible issue and not hard to fix, just one we need to discuss as a group :) 

Other troubles I ran into:
- Making sure the colormap type and range are set *outside* the processInputData function, because they have to stay the same for all facets.
- How to represent no data from plot.data's perspective. See the PR mentioned above for details. Basically there will be one new row for no data, containing x, y as usual, then NaN for overlay.

Leaving the following features to future cards:
- User-defined colorscale type, min, max, midpoint
- passing dates to overlay
- Tooltip show correct name and overlay value

### CODE REVIEW

**Comment by bobular** [comment]
Hi @asizemore - is there any way I can test this without running my own data service?

**Comment by asizemore** [comment]
Hey @bobular! Originally i thought no but i have one idea to try. For background, the data service change is needed because right now it's preventing anything with more than 8 values from being selected as a variable. Additionally, plot.data needs the overlay var to be continuous to make it format the data correctly and send back the seriesGradientColorscale column. 

So, if we can find a very particular variable that is dataShape=Continuous and has 8 or fewer values, we can use that variable to test! I'll look around to see if i can find such a var and let you know! 

If we can't find the magic variable and we go the local backend route, i've had good luck lately running Ryan's eda-local-backend that he walks through in the developer help doc. With the local backend, in the RServe docker file the plot.data installation should be changed from whatever version is there to the branch `gradient-colorscale-nodata`, and the data service should be on the `allow-continuous-overlays` branch. And then note we need the `graident-colormap` branch of web-components as well. 

I'm sorry it's so much to test this one PR! I'll go on a hunt for a special var that can hopefully save us from all of this!

**Comment by bobular** [comment]
Hi @asizemore - oh yes, I'd forgotten there was Ryan's helpful contribution since I last tried this.

I will try that tomorrow.

**Comment by asizemore** [comment]
@bobular I also just talked with @d-callan about the trouble of having to stand up the whole backend just to test this PR. She suggested I could just merge the backend changes first, that way it'd be much, much easier to test and review in the front end. 

It does mean that during the time between backend changes being deployed and these web-eda changes being merged/deployed, we'll have the ability to choose continuous vars as overlays but if one does choose such a var the viz will err. Not the worst. 

Long story short, why don't I try to get the back end PRs merged asap and then get back to you when they're done?

File: src/lib/core/components/visualizations/InputVariables.tsx [hunk 1]
```diff
@@ -137,10 +135,37 @@ export function InputVariables(props: Props) {
...
+              .reduce((disabledVarArray1, disabledVarArray2) =>
+                disabledVarArray1.filter((disabledVar1) =>
+                  disabledVarArray2.some(
+                    (disabledVar2) =>
+                      disabledVar1.variableId === disabledVar2.variableId &&
+                      disabledVar1.entityId === disabledVar2.entityId
+                  )
+                )
+              )
+          : [{ entityId: '', variableId: '' }]; // Typescript help please!!! I tried [{}] as DataElementConstraintRecord but no luck :/
```
Comment (dmfalke) [inline src/lib/core/components/visualizations/InputVariables.tsx:167]: I would take a different approach. Instead of modifying this logic, I would modify `excludedVariables` so that it takes an array of `DataElementConstraint`:

```typescript
export function excludedVariables(
  rootEntity: StudyEntity,
  constraints?: DataElementConstraint[]
): VariableDescriptor[]
```

I think it lead to more flexible code in the long run.

File: src/lib/core/components/visualizations/InputVariables.tsx [hunk 2]
```diff
@@ -239,7 +264,7 @@ export function InputVariables(props: Props) {
                   rootEntity={entities[0]}
                   disabledVariables={disabledVariablesByInputName[input.name]}
                   customDisabledVariableMessage={
-                    flattenedConstraints?.[input.name].description
+                    constraints?.[0][input.name].description // just take first description for now
```
Comment (dmfalke) [inline src/lib/core/components/visualizations/InputVariables.tsx:267]: This code could potentially result in a reference error, e.g., if the array is empty. It might be worth making a helper function to extract the description and to make it a little more defensive. This is an edge case that typescript doesn't handle very well, unfortunately.

**Review by dmfalke — COMMENTED** [review]
@asizemore I have reviewed the code dealing with changes to variable constraints. Overall, it looks great. I did add some suggestions. Let me know if you have any questions!

**Comment by bobular** [comment]
Hi @asizemore - is the lazy option to test this available yet?

**Comment by asizemore** [comment]
Hey @bobular! No not yet - we had some issues with people getting sick :( But thank you for checking in!

I actually talked with Danielle about it this morning. Since there's not much dev time left between now and the b56 release (i'm taking a week off in early feb), we're planning to just pause gradient colormap and wait to merge the lazy test option until after b56. The gradient colormap will likely need lots of feedback from Danica, Sheena, and Dan, before really feeling confident that the it can go out, and since it's not a requirement for clinepi's b56, seems reasonable to pause this effort so that we can focus on getting other requirements done!

I'll bring this up in the meeting tomorrow as well just so everyone knows and see if that plan makes sense to others as well!

**Comment by chowington** [comment]
This branch has been brought up-to-date in https://github.com/VEuPathDB/web-eda/pull/1455
